components/Controls.tsx

```diff
@@ -70,8 +70,21 @@ function BikeStats({ bike }: { bike: Bike }) {
     )
 }
 
+function isNewerFirmwareVersion (newVer: string, oldVer: string) {
+    const oldParts = oldVer.split('.')
+    const newParts = newVer.split('.')
+    for (var i = 0; i < newParts.length; i++) {
+        const a = ~~newParts[i] // parse int
+        const b = ~~oldParts[i] // parse int
+        if (a > b) return true
+        if (a < b) return false
+    }
+    return false
+}
+
 function SpeedLimit({ bike }: { bike: Bike }) {
     const [currentSpeedLimit, setCurrentSpeedLimit] = useState<SpeedLimitEnum | undefined>(undefined)
+    const [firmwareVersion, setFirmwareVersion] = useState<string | undefined>(undefined)
```

Comment:
How about we read the speed limit already in the bike authentication method and then here we can read a public field from the Bike class.